cli/cli-extensions.md

```diff
@@ -0,0 +1,195 @@
+
+And the first one found will be executed with all arguments. And since underscores and dashes are the same,
+`temporal-foo-bar_baz` executable is called for both `temporal foo bar-baz qux` and `temporal foo bar_baz qux` though
+help text only shows the former.
+
+Built-in commands cannot be overridden by extensions, but subcommands can be added to existing commands.
+
+### Discovery and Help Text
+
+Running `temporal --help` (or `temporal help`) _does not_ list extensions. But running `temporal help -a` (or `--all`)
+_does_ list all extensions by traversing the `PATH` and getting every `temporal-`-prefixed executable. There is no short
+description for any of these extensions, they are simply shown as available external commands. All extensions on the
+`PATH` are shown as they are, even if they are multiple commands deep. For example `temporal-foo-bar` and
+`temporal-workflow-dosomething-somethingelse` are shown as `foo bar` and `workflow dosomething somethingelse` in the
+list respectively, not as just `foo` or implied as part of the `workflow` command.
+
+Extensions on built-in commands are not shown as part of the parent command's help.
+
+Running `temporal help <command sequence>` is treated as `temporal <command sequence> --help`. So the same style of
+lookup is performed for extensions.
+
+### Invocation and Flags
+
+Temporal CLI supports arbitrarily ordered flags. So `temporal --address foo workflow start`,
+`temporal workflow --address foo start`, and `temporal workflow start --address foo` are all the same. It is still a
+subject of active research on whether flags of _parent_ commands can still be placed before the extension command in the
+CLI. Regardless, flags of the extension command _must_ come after the extension command because CLI does not know
+whether a flag takes an argument or not so it cannot disambiguate. For example, CLI
+`temporal --myflag someval1 someval2` cannot determine whether that is `temporal someval1 someval2 --myflag` or
+`temporal someval2 --myflag someval1`. So the latter forms must be the forms used for extension-specific flags.
+
+All flags of the parent command _should_ be handled properly by the extension. This means even root-level flags. So for
+example, `temporal-foo` _should_ handle root-level flags like `--output`. However, this isn't a _must_ requirement. See
+the helper library section for some helpers.
+
+Currently, the only root-level flag that the parent `temporal` process respects when calling an extension is
+`--command-timeout` (even though it is still passed along). All other root-level flags are up to the extension to
+handle.
+
+Invocation of the extension is done as a subprocess. Stdin, stdout, stderr, exit code, etc are all handled by the
+extension and just relayed through the `temporal` process as is. It is an area of active research on whether interrupt
+signals can be ignored by the parent `temporal` process to be handled by the subprocess, though that is the goal.
```

Comment:
exec is an option on non-Windows platforms -- that way signals, exit code etc all work exactly as expected without any work. But of course subprocess has to be supported for Windows, so there may not be a justification for the two paths.

Reply:
Yes, we may take this approach as this is what `kubectl` does: https://github.com/kubernetes/kubectl/blob/285ed6ce48d8ba81d430845306c70533170cab58/pkg/cmd/cmd.go#L239-L259. The problem is that I don't think we can do anything as a wrapping process anymore, like enforce `command-timeout`, if we use this form of process replacement. But it still may be worth it.